Generalize CheckpointWrapper

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83055
* __->__ #83035
* #82892

Allow checkpoint_wrapper to take in the checkpoint functional impl. This decouples it from torch.utils.checkpoint and allows other checkpoint implementations to be used.

rohan-varma committed:
Generalize CheckpointWrapper

[ghstack-poisoned]

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -124,7 +143,10 @@ def checkpoint_wrapper(
             The checkpointing implementation to use. Currently only
-            CheckpointImpl.REENTRANT is supported.
+            CheckpointImpl.REENTRANT is supported. Note that this will only
+            be passed into the `torch.utils.checkpoint.checkpoint`
```
nit: I think we need double backtick for it to render as code type since this is a public API.

\`\`torch.utils.checkpoint.checkpoint\`\`

awgu replied:
Just as a probing question: Is the doc that says "Currently only CheckpointImpl.REENTRANT is supported." accurate still?

rohan-varma replied:
Nope, we should update all this documentation.

awgu commented on test/distributed/fsdp/test_checkpoint_wrapper.py:
```diff
@@ -27,7 +27,12 @@ def setUp(self):
+        lin = checkpoint_wrapper(
+            lin,
+            checkpoint_fn=checkpoint,
+            # checkpoint kwargs
```
Is there any case we want to use the `*checkpoint_fn_args` at the moment?

rohan-varma replied:
Not at the moment.

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -124,7 +143,10 @@ def checkpoint_wrapper(
+            specified.
         offload_to_cpu (Optional[bool]):
             Whether to offload outer activations to CPU. Note that this
             currently only works with CheckpointImpl.REENTRANT.
```
Should we add docs for the new arguments `checkpoint_fn`, `checkpoint_fn_args`, and `checkpoint_fn_kwargs`?

rohan-varma replied:
Yes, great catch!

rohan-varma committed:
Update on "Generalize CheckpointWrapper"


Allow checkpoint_wrapper to take in the checkpoint functional impl. This decouples it from torch.utils.checkpoint and allows other checkpoint implementations to be used.

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge

awgu commented on torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py:
```diff
@@ -144,12 +144,18 @@ def checkpoint_wrapper(
+            Functional checkpoint implementation to use. If this is specified,
+            it will be used over the default ``torch.utils.checkpoint.checkpoint``
+            implementation and the `checkpoint_impl` argument will be ignored.
+        *checkpoint_fn_args: (Sequence[Any]): Arguments to pass into `checkpoint_fn`.
```
nit: We might need to double backtick these \`checkpoint_fn\` too (in a follow-up).
